### PR TITLE
[community] Add timeout control and retry for UC tool execution

### PR DESCRIPTION
**Pull request by serena-ruan**
Add timeout at client side for UCFunctionToolkit and add retry logic.
Users could specify environment variable `UC_TOOL_CLIENT_EXECUTION_TIMEOUT` to increase the timeout value for retrying to get the execution response if the status is pending. Default timeout value is 120s.

- [ ] **Add tests and docs**: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

Tested in Databricks:
image

- [x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

**Comment by serena-ruan** [comment]
@harupy Could you stamp this PR as well as you reviewed my last one :D Thanks!

File: libs/community/langchain_community/tools/databricks/_execution.py [hunk 1]
```diff
@@ -174,13 +180,40 @@ def execute_function(
...
+        client_execution_timeout = int(
+            os.environ.get(
+                UC_TOOL_CLIENT_EXECUTION_TIMEOUT,
+                DEFAULT_UC_TOOL_CLIENT_EXECUTION_TIMEOUT,
+            )
+        )
+        while wait_time < client_execution_timeout:
+            wait = min(2**retry_cnt, client_execution_timeout - wait_time)
+            time.sleep(wait)
+            _logger.info(f"Retry times: {retry_cnt}")
```
Comment (BenWilson2) [inline libs/community/langchain_community/tools/databricks/_execution.py:197]: Can we change this to a debug statement?

File: libs/community/langchain_community/tools/databricks/_execution.py [hunk 2]
```diff
@@ -174,13 +180,40 @@ def execute_function(
         parameters=parametrized_statement.parameters,
         **execute_statement_args,  # type: ignore
     )
-    status = response.status
-    assert status is not None, f"Statement execution failed: {response}"
-    if status.state != StatementState.SUCCEEDED:
-        error = status.error
+    if response.status and job_pending(response.status.state) and response.statement_id:
+        statement_id = response.statement_id
+        _logger.info("Retrying to get statement execution status...")
```
Comment (BenWilson2) [inline libs/community/langchain_community/tools/databricks/_execution.py:185]: Can we merge this statement with the counter and add the iteration count + retry delay within the statement to make the informative statement more useful?

File: libs/community/langchain_community/tools/databricks/_execution.py [hunk 3]
```diff
@@ -174,13 +180,40 @@ def execute_function(
...
+                UC_TOOL_CLIENT_EXECUTION_TIMEOUT,
+                DEFAULT_UC_TOOL_CLIENT_EXECUTION_TIMEOUT,
+            )
+        )
+        while wait_time < client_execution_timeout:
+            wait = min(2**retry_cnt, client_execution_timeout - wait_time)
+            time.sleep(wait)
+            _logger.info(f"Retry times: {retry_cnt}")
+            response = ws.statement_execution.get_statement(statement_id)  # type: ignore
+            if response.status is None or not job_pending(response.status.state):
```
Comment (BenWilson2) [inline libs/community/langchain_community/tools/databricks/_execution.py:199]: Do we also want a circuit breaker here if the response returns as a complete to break out of the while loop?

if the response.status.state is StatementState.SUCCEEDED, we should break and resume, right?

Comment (serena-ruan) [reply libs/community/langchain_community/tools/databricks/_execution.py:199]: That is `not job_pending(response.status.state)` right? we just break

**Comment by efriis** [comment]
converting to draft - let me know when it's ready for a review on the langchain side!

As I've mentioned on a few other PRs, it would make sense to migrate the UC tools to the langchain-databricks repo.

**Comment by serena-ruan** [comment]
> converting to draft - let me know when it's ready for a review on the langchain side!
> 
> As I've mentioned on a few other PRs, it would make sense to migrate the UC tools to the langchain-databricks repo.

Thanks, we'll have some new implementation for langchain UC tools support, so want to keep the old one here :)

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Comment by BenWilson2** [comment]
@efriis Hey Erick! :) This one's ready for your review. Thanks!

**Comment by serena-ruan** [comment]
Hi @efriis Could we merge this to catch the next release? Thanks!

**Comment by serena-ruan** [comment]
Hi @baskaryan Could you help merge this PR? Thanks!

**Review by efriis — APPROVED** [review]

**Comment by efriis** [comment]
Sorry for the delay!

**Pull request merged**